core/modules/replica/ControlThread.h

```diff
@@ -0,0 +1,368 @@
+#include <memory>
+#include <string>
+#include <vector>
+
+// Qserv headers
+#include "replica/Job.h"
+#include "replica/Controller.h"
+#include "replica/ServiceProvider.h"
+#include "util/BlockPost.h"
+#include "util/Issue.h"
+#include "util/Mutex.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+// This header declarations
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+/**
+ * Class ControlThreadError represents exceptions throw by the method
+ * of class ControlThread on various error conditions.
+ */
+class ControlThreadError
+    :   public util::Issue {
+public:
+    ControlThreadError(util::Issue::Context const& ctx,
+                       std::string const& message);
+};
+
+/**
+ * Class ControlThreadStopped represents exceptions throw by subclasses
+ * (or method invoked by subclasses) when running subclass-specific
+ * activities as a response to activity cancellation requests. Note, that
+ * this kind of exception is not considered an error.
+ */
+class ControlThreadStopped
+    :   public std::runtime_error {
+public:
+    ControlThreadStopped()
+        :   std::runtime_error("thread stopped") {
+    }
+};
+
+/**
+ * Class ControlThread is the base class for the Controller-side activities run within
+ * dedicated threads. The class provides a public interface for starting and stopping
+ * the activities, notifying clients on abnormal termination of the activities,
+ * as well as an infrastructure supporting an implementation of these activities.
+ */
+class ControlThread
+    :   public std::enable_shared_from_this<ControlThread> {
+
+public:
+
+    /// The pointer type for instances of the class
+    typedef std::shared_ptr<ControlThread> Ptr;
+
+    /// The function type for notifications on the abnormal termination of the threads
+    typedef std::function<void(Ptr)> CallbackType;
+
+    /// The function type for functions used in evaluating user-defined early-termination
+    /// conditions for aborting thread completion tracking
+    typedef std::function<bool(Ptr)> WaitEvaluatorType;
+
+    // Copy semantics is prohibited
+
+    ControlThread(ControlThread const&) = delete;
+    ControlThread& operator=(ControlThread const&) = delete;
+
+    virtual ~ControlThread() = default;
+
+    /// @return a reference to a provider of services
+    ServiceProvider::Ptr const& serviceProvider() const { return _controller->serviceProvider(); }
+
+    /// @return a reference to the Controller
+    Controller::Ptr const& controller() const { return _controller; }
+
+    /// @return the name of the thread
+    std::string name() const { return _name; }
+
+    // @return 'true' if the thread is running
+    bool isRunning() const { return _isRunning.load(); }
+
+    /**
+     * Start a subclass-supplied sequence of actions (virtual method 'run')
+     * within a new thread if it's not running. Note that the lifetime of
+     * the current object is guaranteed to be extended for the duration of
+     * the thread.
```

Comment:
what is 'the current object'? Is it the instance of this class? I don't see how the thread guarantees this object will be kept alive. Can you describe it to me please?

Reply:
This class inherits from:
```
class ControlThread
    :   public std::enable_shared_from_this<ControlThread> {
```
An object of this class is owned by a shared pointer:
```
    typedef std::shared_ptr<ControlThread> Ptr;
```
When class's method `start` is called this method launches `std::thread` and passes a copy of the smart pointer `ControlThread::Ptr` to the object (self). This would guarantee that an instance of `ControlThread` will stay around for a duration of the launched thread (`std::thread`). Therefore it's safe for `std::thread` to call back an instance of `ControlThread`.

Reply:
improved the documentation